packages/flutter_tools/lib/src/test/integration_test_device.dart

```diff
@@ -0,0 +1,140 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+import 'package:stream_channel/stream_channel.dart';
+import 'package:vm_service/vm_service.dart' as vm_service;
+
+import '../application_package.dart';
+import '../base/common.dart';
+import '../build_info.dart';
+import '../device.dart';
+import '../globals.dart' as globals;
+import '../vmservice.dart';
+import 'test_device.dart';
+
+const String kIntegrationTestExtension = 'Flutter.IntegrationTest';
+const String kIntegrationTestData = 'data';
+const String kIntegrationTestMethod = 'ext.flutter.integrationTest';
+
+class IntegrationTestTestDevice implements TestDevice {
+  IntegrationTestTestDevice({
+    @required this.id,
+    @required this.device,
+    @required this.debuggingOptions,
+    @required this.userIdentifier,
+  });
+
+  final int id;
+  final Device device;
+  final DebuggingOptions debuggingOptions;
+  final String userIdentifier;
+
+  ApplicationPackage _applicationPackage;
+  final Completer<void> _finished = Completer<void>();
+  final Completer<Uri> _gotProcessObservatoryUri = Completer<Uri>();
+
+  /// Starts the device.
+  ///
+  /// [entrypointPath] must be a path to an uncompiled source file.
+  @override
+  Future<StreamChannel<String>> start(String entrypointPath) async {
+    final TargetPlatform targetPlatform = await device.targetPlatform;
```

Comment:
Yes, I think we could make `flutter_platform.dart` keep track of some sort of mutable state where rebuilding is skipped for subsequent workflows. For the first run we basically do what is here in this PR, and for subsequent runs we would invoke portions of the attach workflow, to attach and then hot restart the application. I can't be sure if it's sufficient without a prototype, but it seems like a `reloadExistingApp` boolean parameter here would be sufficient. It's also fine if we want to make breaking changes to the interface here because it's all internal anyway.

Reply:
Sounds reasonable to me